openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,358 @@
+
+    return callbackRef;
+}
+
+/**
+ * Releases the OCSP callback.
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, freeOcspCallback)(TCN_STDARGS, jlong ctx, jobject callback) {
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    netty_ocsp_free_callback(e, c, callback);
+}
+
+/**
+ * Enables or disables OCSP stapling for the given SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setStatusType)(TCN_STDARGS, jlong ssl, jint type) {
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+
+    long error = SSL_set_tlsext_status_type(ssl_, type);
+    return (error == 1L) ? JNI_TRUE : JNI_FALSE;
+}
+
+/**
+ * Server: Sets the stapled OCSP record bytes for the given SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setOcspStaple)(TCN_STDARGS, jlong ssl, jbyteArray response) {
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+
+    jsize length = (*e)->GetArrayLength(e, response);
+    if (length == 0) {
+        return JNI_FALSE;
+    }
+
+    //
+    // ATTENTION: This took a while to figure out but OpenSSL wants to (and will)
+    // free() this pointer on its own. Give it something it can free or it will crash.
+    //
+    unsigned char *value = OPENSSL_malloc(sizeof(unsigned char) * length);
+    if (value == NULL) {
+        tcn_ThrowException(e, "OPENSSL_malloc() returned null");
+        return JNI_FALSE;
+    }
+
+    (*e)->GetByteArrayRegion(e, response, 0, length, (jbyte*)value);
+
+    long error = SSL_set_tlsext_status_ocsp_resp(ssl_, value, length);
+    if (error != 1L) {
+        OPENSSL_free(value);
+        return JNI_FALSE;
+    }
+
+    return JNI_TRUE;
+}
+
+/**
+ * Client: Returns the stapled OCSP record bytes as sent by the server or null
+ * if the server didn't provide an OCSP staple.
+ */
+TCN_IMPLEMENT_CALL(jbyteArray, SSL, getOcspStaple)(TCN_STDARGS, jlong ssl) {
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return NULL;
+    }
+
+    unsigned char *response;
+    jint length = (jint)SSL_get_tlsext_status_ocsp_resp(ssl_, &response);
+    if (length < 0) {
+        return NULL;
+    }
+
+    jbyteArray value = (*e)->NewByteArray(e, length);
+    (*e)->SetByteArrayRegion(e, value, 0, length, (jbyte*)response);
+    return value;
+}
+
+jint netty_ocsp_OnLoad(JNIEnv *env) {
+    if (OCSP_CALLBACK_CLAZZ != NULL) {
+        return JNI_ERR;
+    }
+
+    jclass clazz = (*env)->FindClass(env, "io/netty/tcnative/jni/ocsp/OcspCallback");
+    if (clazz == NULL) {
+        return JNI_ERR;
+    }
+
+    OCSP_CALLBACK_CLAZZ = (*env)->NewGlobalRef(env, clazz);
+    if (OCSP_CALLBACK_CLAZZ == NULL) {
+        return JNI_ERR;
+    }
+
+    OCSP_CALLBACK_METHOD = (*env)->GetMethodID(env, OCSP_CALLBACK_CLAZZ, "callback", "(J)I");
+    if (OCSP_CALLBACK_METHOD == NULL) {
+        return JNI_ERR;
```

Comment:
thanks for clarifying ... we rely upon an exception pending and being popped off when we transition from JNI to java.